File: BE/src/main/java/com/fpt/evcare/service/PaymentService.java
package com.fpt.evcare.service;

import com.fpt.evcare.dto.request.payment.CreatePaymentRequest;
import com.fpt.evcare.dto.response.PaymentResponse;
import com.fpt.evcare.dto.response.PageResponse;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

public interface PaymentService {
    
    /**
     * Tạo payment URL từ VNPay hoặc MoMo
     */
    PaymentResponse createPayment(CreatePaymentRequest request);
    
    /**
     * Xử lý callback từ VNPay sau khi thanh toán
     */
    boolean handleVnpayCallback(String vnp_Amount, String vnp_BankCode, String vnp_BankTranNo,
                               String vnp_CardType, String vnp_OrderInfo, String vnp_PayDate,
                               String vnp_ResponseCode, String vnp_TmnCode, String vnp_TransactionNo,
                               String vnp_TransactionStatus, String vnp_TxnRef, String vnp_SecureHash);
    
    /**
     * Xử lý callback từ MoMo sau khi thanh toán
     */
    boolean handleMomoCallback(String partnerCode, String orderId, String requestId,
                               String amount, String resultCode, String transId,
                               String payType, String signature);
    
    /**
     * Kiểm tra trạng thái thanh toán
     */
    PaymentResponse checkPaymentStatus(UUID transactionId);
    
    /**
     * Lấy lịch sử giao dịch theo invoice
     */
    PageResponse<PaymentResponse> getPaymentHistoryByInvoice(UUID invoiceId, Pageable pageable);
    
    /**
     * Lấy lịch sử giao dịch
     */
    PageResponse<PaymentResponse> getAllPaymentHistory(Pageable pageable);
}

File: FE/src/service/paymentService.ts
import { apiClient } from './api';
import type { 
  CreatePaymentRequest, 
  PaymentStatusResponse,
  PagePaymentResponse 
} from '../types/payment.types';

const PAYMENT_API_PATH = '/api/payment';

export const paymentService = {
  /**
   * Tạo payment URL
   */
  async createPayment(request: CreatePaymentRequest): Promise<PaymentStatusResponse> {
    const response = await apiClient.post<PaymentStatusResponse>(
      `${PAYMENT_API_PATH}/create`,
      request
    );
    return response.data;
  },

  /**
   * Kiểm tra trạng thái thanh toán
   */
  async checkPaymentStatus(transactionId: string): Promise<PaymentStatusResponse> {
    const response = await apiClient.get<PaymentStatusResponse>(
      `${PAYMENT_API_PATH}/status/${transactionId}`
    );
    return response.data;
  },

  /**
   * Lấy lịch sử thanh toán theo invoice
   */
  async getPaymentHistoryByInvoice(
    invoiceId: string,
    page: number = 0,
    size: number = 10,
    sortBy: string = 'createdAt',
    sortDir: string = 'DESC'
  ): Promise<PagePaymentResponse> {
    const response = await apiClient.get<PagePaymentResponse>(
      `${PAYMENT_API_PATH}/invoice/${invoiceId}`,
      {
        params: { page, size, sortBy, sortDir }
      }
    );
    return response.data;
  },

  /**
   * Lấy toàn bộ lịch sử thanh toán
   */
  async getAllPaymentHistory(
    page: number = 0,
    size: number = 10,
    sortBy: string = 'createdAt',
    sortDir: string = 'DESC'
  ): Promise<PagePaymentResponse> {
    const response = await apiClient.get<PagePaymentResponse>(
      `${PAYMENT_API_PATH}/history`,
      {
        params: { page, size, sortBy, sortDir }
      }
    );
    return response.data;
  },

  /**
   * Redirect đến payment URL
   */
  redirectToPayment(paymentUrl: string): void {
    window.location.href = paymentUrl;
  },
};
